src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -78,27 +104,69 @@ etaConversionPass ast = do
 -- Function declarations                                                     --
 -------------------------------------------------------------------------------
 
--- | Applies 'etaConvertExpr' to the right-hand side of the given function
---   declaration.
+-- | Applies 'etaConvertTopLevelRhs' to the right-hand side of the given function
+--   declaration to ensure all functions and constructors on the right-hand side 
+--   are fully applied. 
+--   The missing top-level arguments are also added to the left-hand side of the 
+--   declaration and the function's type and the environment are updated accordingly. 
 etaConvertFuncDecl :: IR.FuncDecl -> Converter IR.FuncDecl
 etaConvertFuncDecl funcDecl = do
-  rhs' <- etaConvertExpr (IR.funcDeclRhs funcDecl)
-  return funcDecl { IR.funcDeclRhs = rhs' }
+  let rhs = IR.funcDeclRhs funcDecl
+  xs          <- generateFullArgumentList rhs
+  returnType' <- inEnv
+    $ lookupReturnType IR.ValueScope (IR.funcDeclQName funcDecl)
+  -- compute the function's new (uncurried) type. Assumes that funcDecl's return type has already been inferred.
+  let (args, returnType) = splitReturnType (length xs) (fromJust returnType')
+  -- compute the function's new arguments and add them to the argument list 
+  let newArgs = zipWith
+        (\t i -> IR.VarPat { IR.varPatSrcSpan = NoSrcSpan
+                           , IR.varPatIdent   = i
+                           , IR.varPatType    = Just t
+                           }
+        )
+        args
+        xs
+  let vars' = IR.funcDeclArgs funcDecl ++ newArgs
+  -- compute the new right-hand side
+  rhs'       <- etaConvertTopLevelRhs newArgs rhs
+  -- update the environment with the new type and arguments
+  Just entry <- inEnv $ lookupEntry IR.ValueScope (IR.funcDeclQName funcDecl)
+  modifyEnv $ addEntry entry { entryArity      = length vars'
```

Comment:
Good catch that the environment entries need to be updated : )

However, when the arity of a function in the environment changes, additional η-conversions become necessary. Therefore, η-conversion has to be performed recursively.